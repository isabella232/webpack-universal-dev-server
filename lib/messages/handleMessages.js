// @flow

import type { ChildProcess } from "child_process";
import type { Message, MessageHandler, MessageHandlers } from "../types";

type ProcessLike = Process | ChildProcess;
type ProcessMessage = Message;
type ProcessMessageHandler = MessageHandler<ProcessMessage>;

const passMessage = (
    message: ProcessMessage,
    handler?: ProcessMessageHandler
): void =>
    void (
        handler && handler(message)
    );
const isValidMessage = (
    message
): boolean =>
    message && typeof message.type === "string";
const handleProcessMessage = (
    process: ProcessLike,
    handlers: MessageHandlers
) => (
    message: ProcessMessage
): void =>
    void (
        isValidMessage(message) &&
        passMessage(
            message,
            handlers.get(message.type)
        )
    );

export default (
    process: ProcessLike,
    handlers: MessageHandlers
): void =>
    void (
        process.on("message", handleProcessMessage(process, handlers))
    );
