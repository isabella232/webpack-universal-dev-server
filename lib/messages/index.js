// @flow

import type {
    WpDevServerPauseMessage, WpDevServerResumeMessage,
    WpCompilationMessage, WpDoneMessage,
    WebpackStats
} from "../types"

export const TYPE_WP_DEV_SERVER_PAUSE = "WP_DEV_SERVER/PAUSE"
export const TYPE_WP_DEV_SERVER_RESUME = "WP_DEV_SERVER/RESUME"
export const TYPE_WP_COMPILATION = "WP/COMPILATION"
export const TYPE_WP_DONE = "WP/DONE"
export const TYPE_APP_SERVER_LISTENING = "APP_SERVER/LISTENING"

export const wpDevServerPause = (
): WpDevServerPauseMessage =>
    ({
        type: TYPE_WP_DEV_SERVER_PAUSE
    })

export const wpDevServerResume = (
): WpDevServerResumeMessage =>
    ({
        type: TYPE_WP_DEV_SERVER_RESUME
    })

export const wpCompilation = (
    publisher: string
): WpCompilationMessage =>
    ({
        type: TYPE_WP_COMPILATION,
        publisher
    })

export const wpDone = (
    publisher: string,
    stats: WebpackStats
): WpDoneMessage =>
    ({
        type: TYPE_WP_DONE,
        publisher,
        duration: stats.endTime - stats.startTime,
        hasErrors: stats.hasErrors(),
        hasWarnings: stats.hasWarnings(),
        errorLog: stats.toString("errors-only")
    })

// appServerListening message is not defined here because the message is created by the application itself
// See also our webpack-universal-tools library
