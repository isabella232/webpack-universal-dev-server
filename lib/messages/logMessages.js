// @flow

import { EOL } from "os";
import chalk from "chalk";

const separator = "---------------------------------------------------------------------------------------";
const finished = (
    what: string
): string =>
    chalk.green(`${ what } finished`);
const formatDuration = (
    duration: number
): string =>
    (duration / 1000).toFixed(1);

export const buildStarted = (
    what: string
): string =>
    `${ chalk.yellow(what) } started...`;

export const buildSuccess = (
    what: string,
    duration: number
): string =>
    `${ finished(what) } after ${ formatDuration(duration) }s`;

export const buildError = (
    what: string,
    error: string
): string =>
    [
        "",
        separator,
        chalk.red(`${ what } error`),
        separator,
        "",
        error,
        "",
        separator,
    ].join(EOL);
