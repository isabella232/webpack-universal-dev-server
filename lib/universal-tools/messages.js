// @flow

import type {
    AppServerListeningMessage,
} from "./types";

export const TYPE_APP_SERVER_LISTENING = "APP_SERVER/LISTENING";

export const appServerListening = (
    port: number,
    host: string = "localhost"
): AppServerListeningMessage =>
    ({
        type: TYPE_APP_SERVER_LISTENING,
        port,
        host,
    });
