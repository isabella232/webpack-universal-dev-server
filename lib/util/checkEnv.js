// @flow

import { WUDSError } from "./error";

type ProcessEnv = { [key: string]: ?string };
type RequiredKeys = Array<string>;

export default (
    env: ProcessEnv,
    requiredKeys: RequiredKeys
): void => {
    void requiredKeys.forEach(key => {
        if (key in env === false) {
            throw new WUDSError(`The required env variable '${ key }' is not defined on process.env: Please specify it when creating the child_process`);
        }
    });
};
