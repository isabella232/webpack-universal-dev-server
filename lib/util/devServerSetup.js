// @flow

import type { WebpackDevServerSetup, ExpressApp } from "../types";
import pauseMiddleware from "../util/pauseMiddleware";

export default function devServerSetup(originalSetup?: WebpackDevServerSetup): WebpackDevServerSetup {
    return (...args) => {
        const app: ExpressApp = args[0];

        app.use(pauseMiddleware);

        return originalSetup ?
            // disabling some linting rules because we just want to preserve the original behavior as close as possible
            originalSetup.apply(this, args) : // eslint-disable-line no-invalid-this
            undefined; // eslint-disable-line no-undefined
    };
}
