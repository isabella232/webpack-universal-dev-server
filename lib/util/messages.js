// @flow

const NAMESPACE = __filename;

// WDS = Webpack Dev Server (watches, compiles and serves the client bundle)
// WWP = Webpack Watch Process (watches and compiles the server bundle)
export const TYPE_WDS_PAUSE = `${ NAMESPACE }/WDS/PAUSE`;
export const TYPE_WDS_RESUME = `${ NAMESPACE }/WDS/RESUME`;
export const TYPE_WWP_COMPILATION = `${ NAMESPACE }/WWP/COMPILATION`;
export const TYPE_WWP_DONE = `${ NAMESPACE }/WWP/DONE`;

export type Message = {
    type: string;
};

export function wdsPause(): Message {
    return {
        type: TYPE_WDS_PAUSE
    };
}

export function wdsResume(): Message {
    return {
        type: TYPE_WDS_RESUME
    };
}

export function wwpCompilation(): Message {
    return {
        type: TYPE_WWP_COMPILATION
    };
}

export function wwpDone(): Message {
    return {
        type: TYPE_WWP_DONE
    };
}
