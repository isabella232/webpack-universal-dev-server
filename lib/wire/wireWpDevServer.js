// @flow

import type { Process } from "flow";
import type { ChildProcess } from "child_process";
import type { Logger, WpDoneMessage } from "../types";
import handleMessages from "../util/handleMessages";
import { buildStarted, buildSuccess, buildError } from "../util/logMessages";
import { TYPE_WP_COMPILATION, TYPE_WP_DONE } from "../util/messages";

type WireWpDevServerArgs = {
    +wpDevServer: ChildProcess,
    +process: Process,
    +log: Logger
};

export default function wireWpDevServer({
        wpDevServer,
        process,
        log
    } : WireWpDevServerArgs): void {
    const messageHandlers = new Map();

    messageHandlers.set(TYPE_WP_COMPILATION, () => {
        log(buildStarted("webpack client build"));
    });
    messageHandlers.set(TYPE_WP_DONE, (message: WpDoneMessage) => {
        if (message.errorLog) {
            log(buildError("webpack client build", message.errorLog));
        } else if (!message.hasErrors) {
            log(buildSuccess("webpack client build", message.duration));
        }
    });

    handleMessages(wpDevServer, messageHandlers);

    // Webpack errors won't reported to stderr. This is just for hard failures.
    wpDevServer.stderr.pipe(process.stderr);
}
