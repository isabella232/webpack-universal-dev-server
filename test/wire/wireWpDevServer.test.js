import { EOL } from "os"; // We split the logs by EOL to make the snapshot tests OS independent.
import test from "ava";
import { WritableProcess, ReadableProcess } from "../helpers/fakeProcesses";
import wireWpDevServer from "../../lib/wire/wireWpDevServer";
import { wpCompilation, wpDone } from "../../lib/util/messages";

function noop() {}

test("should log a build started message when the wpDevServer process emitted a wpCompilation message", t => {
    const wpDevServer = new ReadableProcess();
    let receivedLog;

    wireWpDevServer({
        wpDevServer,
        process,
        log(log) {
            receivedLog = log;
        }
    });

    wpDevServer.emit("message", wpCompilation("test-publisher"));

    t.snapshot(receivedLog.split(EOL));
});

test("should log a build success message when the wpDevServer process emitted a wpDone message with hasErrors false", t => {
    const wpDevServer = new ReadableProcess();
    const fakeWpStats = {
        startTime: 1000, // fake timestamps, minimum precesion in milliseconds
        endTime: 2500,
        hasErrors() { return false; },
        hasWarnings: noop,
        toString: noop
    };
    let receivedLog;

    wireWpDevServer({
        wpDevServer,
        process,
        log(log) {
            receivedLog = log;
        }
    });

    wpDevServer.emit("message", wpDone("test-publisher", fakeWpStats));

    t.snapshot(receivedLog.split(EOL));
});

test("should log a build error message when the wpDevServer process emitted a wpDone message with an errorLog", t => {
    const wpDevServer = new ReadableProcess();
    const fakeWpStats = {
        hasErrors() { return true; },
        hasWarnings: noop,
        toString() {
            return "Module build failed";
        }
    };
    let receivedLog;

    wireWpDevServer({
        wpDevServer,
        process,
        log(log) {
            receivedLog = log;
        }
    });

    wpDevServer.emit("message", wpDone("test-publisher", fakeWpStats));

    t.snapshot(receivedLog.split(EOL));
});

test("should pipe stderr from wpDevServer to process", async t => {
    const wpDevServer = new ReadableProcess();
    const process = new WritableProcess();

    wireWpDevServer({
        wpDevServer,
        process,
        log: noop
    });

    await new Promise((resolve, reject) => {
        wpDevServer.stderr.on("readable", resolve);
        wpDevServer.stderr.put("test", "utf8");
        wpDevServer.stderr.stop();
    });

    t.is(process.stderr.getContentsAsString("utf8"), "test");
});
