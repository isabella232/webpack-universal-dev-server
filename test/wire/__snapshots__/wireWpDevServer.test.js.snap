exports[`should log a build error message when the wpDevServer process emitted a wpDone message with an errorLog 1`] = `"["","---------------------------------------------------------------------------------------","webpack client build error","---------------------------------------------------------------------------------------","","Module build failed","","---------------------------------------------------------------------------------------"]"`;

exports[`should log a build started message when the wpDevServer process emitted a wpCompilation message 1`] = `"[\"webpack client build started...\"]"`;

exports[`should log a build success message when the wpDevServer process emitted a wpDone message with hasErrors false 1`] = `"[\"webpack client build finished after 1.5s\"]"`;
